index.js

```diff
@@ -110,66 +78,102 @@ function fastifyCors (fastify, opts, next) {
         )
       }
 
-      next()
+      return next()
     })
   }
 
-  function configureOrigin (req, reply, callback) {
-    const reqOrigin = req.headers.origin
-    if (isOriginFunction) {
-      const result = origin.call(fastify, reqOrigin, _onOrigin)
-      if (result && typeof result.then === 'function') {
-        result.then(res => _onOrigin(null, res), callback)
-      }
-    } else {
-      _configureOrigin(origin)
+  function preflightHandler (req, reply) {
+    // Do not handle preflight requests if the origin was not allowed
+    if (!req.corsOriginAllowed) {
+      reply.code(404).type('text/plain').send('Not Found')
+      return
     }
 
-    function _onOrigin (err, origin) {
-      if (err !== null || origin === false) {
-        return callback(err, origin)
-      }
-
-      _configureOrigin(origin)
+    // Strict mode enforces the required headers for preflight
+    if (strictPreflight === true && (!req.headers.origin || !req.headers['access-control-request-method'])) {
+      reply.status(400).type('text/plain').send('Invalid Preflight Request')
+      return
     }
 
-    function _configureOrigin (origin) {
-      if (!origin || origin === '*') {
-        // allow any origin
-        reply.header('Access-Control-Allow-Origin', '*')
-      } else if (isOriginString) {
-        // fixed origin
-        reply.header('Access-Control-Allow-Origin', origin)
-      } else {
-        // reflect origin
-        reply.header(
-          'Access-Control-Allow-Origin',
-          isOriginAllowed(reqOrigin, origin) ? reqOrigin : false
-        )
+    // Handle preflight headers
+    reply.header(
+      'Access-Control-Allow-Methods',
+      Array.isArray(methods) ? methods.join(', ') : methods
+    )
+
+    if (allowedHeaders === null) {
+      vary(reply, 'Access-Control-Request-Headers')
+      const reqAllowedHeaders = req.headers['access-control-request-headers']
+      if (reqAllowedHeaders !== undefined) {
+        reply.header('Access-Control-Allow-Headers', reqAllowedHeaders)
       }
+    } else {
+      reply.header(
+        'Access-Control-Allow-Headers',
+        Array.isArray(allowedHeaders) ? allowedHeaders.join(', ') : allowedHeaders
+      )
+    }
 
-      callback(null, origin)
+    if (maxAge !== null) {
+      reply.header('Access-Control-Max-Age', String(maxAge))
     }
+
+    // Safari (and potentially other browsers) need content-length 0,
+    // for 204 or they just hang waiting for a body
+    reply
+      .code(optionsSuccessStatus)
+      .header('Content-Length', '0')
+      .send()
   }
 
-  function isOriginAllowed (reqOrigin, origin) {
-    if (Array.isArray(origin)) {
-      for (let i = 0; i < origin.length; ++i) {
-        if (isOriginAllowed(reqOrigin, origin[i])) {
-          return true
+  async function resolveOriginWrapper (req, cb) {
```

Comment:
This is _really bad_ and it can cause huge amount of problems. Never mix callbacks callbacks and async.